### ISSUE TITLE
Ensure that we check the YAML syntax of license data files in tests.

### ISSUE THREAD

**Issue by lambdakilo** [nexB/scancode-toolkit#3947]
it is difficult for humans to predict how a given yaml document will parse [1]. For instance https://github.com/aboutcode-org/scancode-licensedb/blob/main/docs/lgpl-2.0-fltk.yml is an invalid yaml document although it doesn't seem like it. I discovered this while parsing the yamls of the repository with pyyaml.

Should we deprecate and remove all of the yaml version from this repo?

1: https://ruudvanasseldonk.com/2023/01/11/the-yaml-document-from-hell

**Comment by pombredanne** [comment]
@akirataguchi115 re:
> Should we deprecate and remove all of the yaml version from this repo?

I agree that YAML can be problematic, and the spec is one of the most complex document that I have seen, so much so that we use yaml frontmatter now for license original data in https://github.com/aboutcode-org/scancode-toolkit/blob/develop/src/licensedcode/data/licenses/3dslicer-1.0.LICENSE

Yet, when used minimally, YAML is highly readable by humans and machines. And it can wrap long strings which makes big difference with JSON when using VCS.

Having the text part of a YAML attribute is likely a mistake, and front matter may be better.

As some folks may rely on this data format, I would rather have tests to validate that the YAML data is correct and can parse alright, rather than dropping it. 

Can I interest you in adding a test? This would be best done in the command line that creates the data at https://github.com/aboutcode-org/scancode-toolkit/blob/develop/src/licensedcode/license_db.py and this could simply fail the data dump. And this would likely fail the test at https://github.com/aboutcode-org/scancode-toolkit/blob/develop/tests/licensedcode/test_license_db.py#L26 that tests the dump.

**Comment by pombredanne** [comment]
TIL: https://www.yaml.info/learn/quote.html#indent via https://github.com/yaml/pyyaml/issues/717#issuecomment-1582649302

**Comment by pombredanne** [comment]
And a solution to get likely correct YAML would be to:
- indent by one the text (though this may be not 100% needed)
- use the "1" notation in `text: |1` to indicate that this is using a one space indent (which is the default unless there are more than one space :weary: )

**Comment by lambdakilo** [comment]
> As some folks may rely on this data format, I would rather have tests to validate that the YAML data is correct and can parse alright, rather than dropping it.

You're right. It seems I was too hasty when I thought dropping YAML as an offered data format would be the best choice.

I would be interested in adding a test :+1: Thanks for asking!

> This would be best done in the command line that creates the data at `/src/licensedcode/license_db.py` and this could simply fail the data dump. And this would likely fail the test at `/tests/licensedcode/test_license_db.py#L26` that tests the dump.

What do the words "data" and "data dump" mean in this context? Shouldn't I create a test in the CI/CD pipeline of `scancode-licensedb` instead so that it would always fail if there exists an invalid yaml file in the repository during a merge or a pull request?

**Comment by pombredanne** [comment]
The data stored here is entirely generated from ScanCode toolkit, so this is the place to fix this.

**Comment by lambdakilo** [comment]
Ah makes sense. I'm actually writing my thesis right now so I might not be able to make the test anytime soon. I'll have a look in my free time though if it would be simple and rewarding enough for me to create the test.

To double make sure I understood it correct: I should create a license dump in YAML format like in lgpl-2.0-fltk.yml in `src/licensedcode/license_db.py` and then I should also modify `tests/licensedcode/test_license_db.py` so that they should fail on a YAML file like the lgpl-2.0-fltkl.yml the way it is now?

**Comment by pombredanne** [comment]
@akirataguchi115 re: your thesis, do not hesitate to reach out for input or feedback! What is the topic?

As for the test, a simple enough approach could be change the code in 
https://github.com/aboutcode-org/scancode-toolkit/blob/9a340fc36b971bcc04fdf255ee73bf88ce39635a/src/licensedcode/license_db.py#L134-L138
from
```Python
        write_file(
            output_path,
            f"{lic.key}.yml",
            saneyaml.dump(license_data_with_text, indent=2)
        )
```
to something more or less like this:
```Python
        as_yaml  = saneyaml.dump(license_data_with_text, indent=2)
        # this will crash loudly if the YAML is invalid
        saneyaml.load(as_yaml)
        write_file(
            output_path,
            f"{lic.key}.yml",
            as_yaml
        )
```

The test suite already attempts to dump all the licenses, so this should fail there too without extra code.

**Comment by lambdakilo** [comment]
> The test suite already attempts to dump all the licenses
Ahh that helps me understand this better. Excited to try out that approach! Thank you.

The topic is public software licenses: https://akirataguchi115.github.io/mscthesis/HY-CS-main.pdf

**Comment by pombredanne** [comment]
The check is already available in https://github.com/aboutcode-org/scancode-toolkit/blob/395091ccfbe4c56285d09d43effaac4619392883/src/licensedcode/models.py#L664 .... but we do not run it may be?

Because the check works:
```
>>> li.validate(thorough=True, licenses=lics)
({'lgpl-2.0-fltk': ['No SPDX license key'], 'GLOBAL': ['Error invalid YAML text at: lgpl-2.0-fltk, failed during saneyaml.load()Run scancode-reindex-licenses --load-dump to fix this.']}, {}, {})
```

BUT note that lgpl-2.0-fltk is deprecated.
I am moving this issues to SCTK.

**Comment by pombredanne** [comment]
We need to make the "thorough" check True in https://github.com/aboutcode-org/scancode-toolkit/blob/395091ccfbe4c56285d09d43effaac4619392883/tests/licensedcode/test_models.py#L89

**Comment by DanielS01ss** [comment]
Hi!
I would like to work on this issue!

**Comment by asimchoudhary** [comment]
**Is this issue has already been resolved?**

In the current codebase, the test in test_models.py sets the thorough parameter to True. As a result, the validation of the YAML file is being performed:
```

errors, warnings, infos = models.License.validate(
    licenses=models.load_licenses(with_deprecated=False),
    verbose=False,
    thorough=True,
)
```

Given this implementation, it appears that the issue has been addressed. Could you confirm if this resolves the issue: "Ensure that we check the YAML syntax of license data files in tests"?

**Comment by tanishapritha** [comment]
Since this issue is marked "open" can I work on this? If that works I'll first send what I plan on doing and will then proceed.

**Comment by Yuganti-hash** [comment]
Hi, I went through the discussion and noticed that YAML validation using the "thorough" flag and saneyaml.load might already be implemented.

Could you please confirm if this issue still needs work or if there are any remaining improvements I can contribute to?

**Comment by Shaurya-dev7** [comment]
Hi, I’ve been working on this issue and explored adding YAML syntax validation during the license dump/test workflow using PyYAML validation checks together with test coverage updates. I’m cleaning up the implementation and will open a PR shortly.
